File: paho-v10/src/main/java/org/eclipse/paho/client/mqttv10/TimerPingSender.java
/*******************************************************************************
 * Copyright (c) 2014 IBM Corp.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * and Eclipse Distribution License v1.0 which accompany this distribution. 
 *
 * The Eclipse Public License is available at 
 *    http://www.eclipse.org/legal/epl-v10.html
 * and the Eclipse Distribution License is available at 
 *   http://www.eclipse.org/org/documents/edl-v10.php.
 */

package org.eclipse.paho.client.mqttv10;

import java.util.Timer;
import java.util.TimerTask;

import org.eclipse.paho.client.mqtt.ILogger;
import org.eclipse.paho.client.mqtt.MqttPingSender;
import org.eclipse.paho.client.mqttv10.internal.ClientComms;


/**
 * Default ping sender implementation
 * 
 * <p>
 * This class implements the {@link IMqttPingSender} pinger interface allowing applications to send ping packet to server every keep alive interval.
 * </p>
 * 
 * @see MqttPingSender
 */
public class TimerPingSender implements MqttPingSender
{
	private ClientComms comms;

	private ILogger logger;

	private Timer timer;

	private final static String className = TimerPingSender.class.getName();

	private final String TAG = "TimerPingSender";

	public void init(ClientComms comms, ILogger logger)
	{
		if (comms == null)
		{
			throw new IllegalArgumentException("ClientComms cannot be null.");
		}
		this.comms = comms;
		this.logger = logger;
	}

	public void start()
	{
		final String methodName = "start";
		String clientid = comms.getClient().getClientId();

		// @Trace 659=start timer for client:{0}

		timer = new Timer("MQTT Ping: " + clientid);
		// Check ping after first keep alive interval.
		timer.schedule(new PingTask(), comms.getKeepAlive());
	}

	public void stop()
	{
		final String methodName = "stop";
		// @Trace 661=stop
		logger.d(TAG, "Stopping timer");
		if (timer != null)
		{
			timer.cancel();
		}
	}

	public void schedule(long delayInMilliseconds)
	{
		timer.schedule(new PingTask(), delayInMilliseconds);
	}

	class PingTask extends TimerTask
	{
		private static final String methodName = "PingTask.run";

		public void run()
		{
			// @Trace 660=Check schedule at {0}
			logger.d(TAG, "in ping timer task run function");
			comms.checkForActivity();
		}
	}
}

File: pingsender/timer-pingsender/src/main/java/com/gojek/timer/pingsender/TimerPingSender.kt
package com.gojek.timer.pingsender

import androidx.annotation.VisibleForTesting
import com.gojek.courier.extensions.fromMillisToSeconds
import com.gojek.courier.extensions.fromNanosToMillis
import com.gojek.courier.logging.ILogger
import com.gojek.courier.utils.Clock
import com.gojek.mqtt.pingsender.IPingSenderEvents
import com.gojek.mqtt.pingsender.MqttPingSender
import com.gojek.mqtt.pingsender.NoOpPingSenderEvents
import com.gojek.mqtt.pingsender.PingActionCallback
import com.gojek.mqtt.pingsender.PingSenderComms
import java.util.Timer
import java.util.TimerTask

/**
 * Default ping sender implementation
 *
 * This class implements the [MqttPingSender] pinger interface allowing applications to send ping packet to server every keep alive interval.
 *
 * @see MqttPingSender
 */
internal class TimerPingSender(
    private val pingSenderConfig: TimerPingSenderConfig,
    private val clock: Clock = Clock(),
    private val timerFactory: TimerFactory = TimerFactory()
) : MqttPingSender {
    private lateinit var comms: PingSenderComms
    private lateinit var logger: ILogger

    @VisibleForTesting
    internal lateinit var timer: Timer

    private var pingSenderEvents: IPingSenderEvents = NoOpPingSenderEvents()

    override fun init(
        comms: PingSenderComms,
        logger: ILogger
    ) {
        this.comms = comms
        this.logger = logger
    }

    override fun start() {
        logger.d(TAG, "Starting timer")
        val clientId = comms.clientId

        timer = timerFactory.getTimer("MQTT Ping: $clientId")
        schedule(comms.keepAliveMillis)
    }

    override fun stop() {
        logger.d(TAG, "Stopping timer")
        if (::timer.isInitialized) {
            timer.cancel()
        }
    }

    override fun schedule(delayInMilliseconds: Long) {
        timer.schedule(PingTask(), delayInMilliseconds)
        pingSenderEvents.mqttPingScheduled(delayInMilliseconds.fromMillisToSeconds(), comms.keepAliveMillis.fromMillisToSeconds())
    }

    override fun setPingEventHandler(pingSenderEvents: IPingSenderEvents) {
        this.pingSenderEvents = pingSenderEvents
    }

    internal inner class PingTask : TimerTask() {
        override fun run() {
            logger.d(TAG, "Sending ping")
            val serverUri = comms.serverURI ?: ""
            val keepAliveMillis = comms.keepAliveMillis
            pingSenderEvents.mqttPingInitiated(serverUri, keepAliveMillis.fromMillisToSeconds())
            val sTime = clock.nanoTime()
            val initiated = comms.checkActivity(
                pingSenderConfig.sendForcePing,
                object : PingActionCallback {
                    override fun onSuccess() {
                        logger.d(TAG, "Mqtt Ping Sent successfully")
                        val timeTaken = (clock.nanoTime() - sTime).fromNanosToMillis()
                        pingSenderEvents.pingEventSuccess(serverUri, timeTaken, keepAliveMillis.fromMillisToSeconds())
                    }

                    override fun onFailure(exception: Throwable) {
                        logger.d(TAG, "Mqtt Ping Sent failed")
                        val timeTaken = (clock.nanoTime() - sTime).fromNanosToMillis()
                        pingSenderEvents.pingEventFailure(
                            serverUri,
                            timeTaken,
                            exception,
                            keepAliveMillis.fromMillisToSeconds()
                        )
                    }
                }
            )
            if (!initiated) {
                logger.d(TAG, "Mqtt Ping Token null")
                pingSenderEvents.pingMqttTokenNull(serverUri, keepAliveMillis.fromMillisToSeconds())
                return
            }
        }
    }
}

private const val TAG = "TimerPingSender"
